### DISCUSSION TITLE
Generate a 512px PNG of a Very Large (8GB) PSB in ~5 sec or less?

### DISCUSSION THREAD

**Discussion by chrismacguy** [ImageMagick/ImageMagick#6987] [Digital Image Processing]
**How can I get this faster (5 seconds or less)?**

The code below creates a 512px image of an **8GB PSB** (multi-layer with lots of Alphas) in about 30 seconds:
`magick 'pathToTest.psb[0]' -resize 512x512^ -gravity center -extent 512x512 -density 72x72 pathOut.png
`

Notes:
* The thumbnail is only 160px which is too small, otherwise, I'd use that (upsizing is not viable).
* PSB is saved with Maximize Compatibility enabled.
* I believe the answer lies in bypassing the raw data and skipping to the composite. 
* Standard Open in Photoshop ~48 sec vs. Composite Data Open in Photoshop (Shift+Option on Mac) ~2 sec.
* My assumption is [0] after the filename bypasses the layered data and just reads the composite, is this correct?
* Without [0] after the filename I get an error _"unable to decompress image `/.../test.psb' @ error/psd.c/ReadPSDChannel/1515"_
* System: MacOS Ventura 13.6.1, MacBook Pro 16" 2019 2.4 GHz 8-Core Intel Core i9, 64 GB Memory, ImageMagick 7.1.1-24

**Comment by snibgo** [comment]
IM doesn't know about layered data and composite data. IM sees a number of images in the file. To read just the first image, use `[0]`. For the second, use `[1]`, etc.

How many pixels are in the composite image?

**Reply by chrismacguy** [reply]
Can you please explain what you mean by the number of images in the file? Does this refer to the layers? 

The test image dimensions are 9620px x 9620px @ 799.382 ppi. It has 409 layers and 25 Alpha Channels.

**Reply by snibgo** [reply]
The command `magick pathToTest.psb info:` will list the images that IM can see in the file, with some information such the dimensions of each image.

**Reply by fmw42** [reply]
How big is your RAM.  You need to consider they pixel dimensions multiplied by the number of layers and and multiplied by the number of channels (including the alpha) times the bit depth per channel.  Do you have enough RAM? Otherwise, you would need to configure IM to use disk to supplement and that would slow it down considerably.  Also see if you have configured your policy.xml file to use all the RAM you have.

**Reply by chrismacguy** [reply]
The command `magick pathToTest.psb info:` took a while to execute but it returned **[74]** 9640x9640 8-b9t sRGB images in the test file.

I am running on a 2.4 Ghz 8-Core Intel  Core i9 MacBook Pro 2019 with 64 GB memory. I haven't done any special configurations to my policy.xml file.

This would need to run from multiple different machines so custom modifications based on each machine probably aren't likely. Was hoping there was some way to quickly get to the composite data that Photoshop stores?

**Reply by fmw42** [reply]
layer [0] would be the composite in the PSB image as Imagemagick would read it..

So

`magick pathToTest.psb[0] -resize 512x512 -density 72x72 out.png`

should give you what you want. Since your input is square and you want a square output, there should be no need to extent, in this case.

**Reply by chrismacguy** [reply]
Thank you for confirming layer [0] is the composite. 

**pathToTest.psb[0]** needs to be quoted (at least on MacOS) to be identified:
`magick 'pathToTest.psb[0]' -resize 512x512 -density 72x72 out.png`

While the Test Image is square, I can't rely on that for all since this could be run on any image, hence the additional steps in my original line of code. 

Even so, the shorter code still took around 30 seconds to complete on my machine. Was hoping for a solution that could get it in under 5 seconds, if any such solution exists...

**Reply by fmw42** [reply]
no quoting for files with [0] appended works fine on my Mac

however if your filename or path include spaces, then quoting would be needed

I believe that Imagemagick has to read the whole file (all layers) to get to layer [0]. So that is why it takes so long.

So if you do not have enough RAM to hold the whole image, then IM will offload to disk, which will increase the time dramatically.  Have you computed the required RAM needed for your image per my comment above? Looks to me like about 3 GB. So that should be OK if your policy is giving your your 64 GB RAM.

**Reply by chrismacguy** [reply]
Odd...there are no spaces in the path but it didn't work without the quotes. Not a big deal though since it's working with them.

I don't know how to compute the RAM needed. I've located the policy.xml. Do I just change the Memory from 256MiB to 3GiB? Anything else? Currently: Map (512MiB), Disk (1GiB), Area (16KP).

**Reply by chrismacguy** [reply]
`magick -list resource` run on my system yields:

Resource limits:
  Width: 36.0288PP
  Height: 36.0288PP
  Area: 274.878GP
  List length: unlimited
  Memory: 64GiB
  Map: 128GiB
  Disk: unlimited
  File: 192
  Thread: 16
  Throttle: 0
  Time: unlimited

**ExifTool** returns the following relevant info on the image:

File Size: 7.9 GB
File Type: PSB
Num Channels: 28
Image Height: 9640
Image Width: 9640
Bit Depth: 8
Color Mode: RGB
X & Y Resolution: 799.38
Alpha Channels: 25
Layer Count: 518
Compression: RLE
Megapixels: 92.9

**Comment by fmw42** [comment]
You could try libvips at https://github.com/libvips/libvips

**Reply by chrismacguy** [reply]
I've never used vips before so would have to figure out how to get what I need from it... I just installed it but when I ran `copy imageIn.psb imageOut.png` it just produced a 48MB white image. I suppose I'd need to figure out how to handle Alpha Channels, asssuming that's what is tripping up the output, unless installing with brew didn't grab everything I needed.

**Reply by fmw42** [reply]
I have given you the formula for computing the required RAM above. But I computed it as about 3 GB. The size of the image file itself does not account for compression, etc.  If you only have 256 MB in your policy, then it is offloading to disk.  Update your policy.xml file for RAM of 6 GB. Also update any other argument such as area, width or height that is lower than in your image.

What do you get from 

`magick -list resource`

**Reply by chrismacguy** [reply]
Resource limits:
Width: 36.0288PP
Height: 36.0288PP
Area: 274.878GP
List length: unlimited
Memory: 64GiB
Map: 128GiB
Disk: unlimited
File: 192
Thread: 16
Throttle: 0
Time: unlimited

**Reply by fmw42** [reply]
I make a mistake above with estimate of 3 GB. 

If we assume you have 9640x9640 dimensions, 4 channels per layer (RGBA), 518 layers, and 8-bits per channel. Then your requirements would be

9640\*9640\*4\*518\*8 = 1,540,401,049,600 bits   or   1540 Gbits or 1540/8 = 192 GBytes

So you do not have enough RAM to hold the image if you only have 64 GB. So it is offloaded to disk.

**Reply by chrismacguy** [reply]
Ahh.. that's a lot of RAM. Too bad it has to read all the layers to get to layer [0]. Assumimg that's what's eating up all the memory?
